Refactor to support all Syslog variants

stevedlawrence opened the pull request:
- Move all Syslog variants into their own files
- Modify the Syslog root so it supports all variants instead of just RFC5424
- Removes portability. Prival is now referenced from multiple schemas so it is difficult to have a single "portable" schema. Instead, instructions and documentation are added describing how make the schema portable.
- Copy Solarwinds tests to Sylog tests with modifications to remove Solarwinds specific parts to ensure the Syslog element supports all variants

mbeckerle commented on README.md:
```diff
@@ -1,32 +1,50 @@
-Removed version number from namespace URI. 
-Changed branding from Tresys to Owl. 
+The common.dfdl.xsd file can be modified to enable support for IBM DFDL. See
+the documentation annotations in the Prival groups for details on the needed
```
I will assume "Prival" is not a mispelling.

stevedlawrence replied:
It's in the spec as "PRIVAL", which looking at the spec is short for Priority Value. I'll update it use the longer name to make it more clear.

mbeckerle commented on README.md:
```diff
@@ -1,32 +1,50 @@
+adjustmens. Additionally, encodingErrorPolicy must be changed from "replace" to
+"error" when used with IBM DFDL.
 
+## Change History
```
I like to call this section Release Notes because we're not trying to recapture everything that is in the git commit log comments, just what is important about the release.

stevedlawrence replied:
Makes sense, I'll update and make sure my additions focus on the more important aspects.

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-cisco.dfdl.xsd:
```diff
@@ -0,0 +1,136 @@
+          <xs:sequence dfdl:separator="%SP;" dfdl:separatorPosition="postfix">
+            <xs:element name="SequenceNumber" type="xs:int" dfdl:textNumberPattern="0" dfdl:terminator=":" />
+            <xs:element name="TimestampMonth" dfdl:lengthKind="explicit" dfdl:length="3">
+              <xs:simpleType>
```
Create shared types.dfdl.xsd and define this reused month3 simple type there.  
Or just use common.dfdl.xsd that you already have.

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-cisco.dfdl.xsd:
```diff
@@ -0,0 +1,136 @@
+            </xs:element>
+            <xs:element name="TimestampTime" type="xs:time" dfdl:calendarPattern="HH:mm:ss" dfdl:lengthKind="explicit" dfdl:length="8" dfdl:terminator=":" />
+            <xs:sequence dfdl:initiator="%%" dfdl:separator="-" dfdl:terminator=":">
+              <xs:element name="Facility" type="xs:string" />
```
Unconstrained string type will fail XSAT-type quality analysis.  Another reason to define shared types is that basically you can never use xs:string except as a base in a simpleType def, if you want to keep XSAT-style checkers happy.

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-cisco.dfdl.xsd:
```diff
@@ -0,0 +1,136 @@
+                  </xs:restriction>
+                </xs:simpleType>
+              </xs:element>
+              <xs:element name="Mnemonic" type="xs:string" />
```
Grep everything for xs:string and replace all usages with your own simpleType with some maxLength facet.

stevedlawrence replied:
The spec doesn't specify max lengths for fields, and different fields would probably have very different lengths. For example, the Cisco Facility field is a short word like "PARSER", but the Message field is going to be significantly longer since it's the actual log message. We could pick a large maxLength of strings like 2048 to make XSAT happy and ensure all log messages pass except those that are clearly excessive, but different environments might know of, and want, tighter restrictions. And for fields like "Facility" that's waay to big of a max length. There's only 20 strings, so doing them individually probably wouldn't be too bad, but I'd basically just be guessing what numbers are reasonable, and could easily be wrong.

stevedlawrence replied:
Actually, RFC5424 does provide maxLengts for some header fields, which we do use. Other formats have similar fields so we can carry over maxLengths for those. The only one that doesn't have a max length is the message, but a large value like 9999 seems reasonable.

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-rfc3164.dfdl.xsd:
```diff
@@ -0,0 +1,130 @@
+            <xs:element name="Hostname">
+              <xs:simpleType>
+                <xs:restriction base="xs:string">
+                  <xs:pattern value="[!-~]+" />
```
Regex with "+" will upset XSAT. Replace with "{1,8}".

stevedlawrence replied:
Hostnames can be much longer than 8 chars, do we just pick a reasonably large number? Or maybe just leave it up to people using these schemas to run XSAT themselves and pick reasonable values?

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-rfc5424.dfdl.xsd:
```diff
@@ -0,0 +1,215 @@
+            <xs:restriction base="xs:string">
+              <xs:minLength value="1" />
+              <xs:maxLength value="255" />
+              <xs:pattern value="[!-~]+" />
```
Replace "+" with "{1,255}". I know it's barfy from an engineering perspective because you have the 255 in multiple places then, but it will silence a warning from an XSAT-style checker that just greps for pattern facets containing * or +.

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-rfc5424.dfdl.xsd:
```diff
@@ -0,0 +1,215 @@
+            <xs:restriction base="xs:string">
+              <xs:minLength value="1" />
+              <xs:maxLength value="48" />
+              <xs:pattern value="[!-~]+" />
```
Yeah, all these + patterns will spit out XSAT-style warnings.

mbeckerle commented on src/main/resources/com/owlcyberdefense/syslog/xsd/syslog-rfc5424.dfdl.xsd:
```diff
@@ -0,0 +1,215 @@
+      <xs:element name="Message" minOccurs="0" dfdl:initiator="%SP;">
+        <xs:complexType>
+          <xs:choice>
+            <xs:element name="UTF8" dfdl:initiator="%#xEF;%#xBB;%#xBF;" type="xs:string" dfdl:encoding="UTF-8" />
```
This is the biggest place where there's an issue with string length. Is there an upper bound here that is part of the spec, that you can put on facets? Suggest 9999 if there isn't some other sensible value.

stevedlawrence replied:
There is no upper bound on length. The RFC5424 spec says:
>  Syslog message size limits are dictated by the syslog transport
   mapping in use.  There is no upper limit per se.  Each transport
   mapping defines the minimum maximum required message length support,
   and the minimum maximum MUST be at least 480 octets in length.
>
>   Any transport receiver MUST be able to accept messages of up to and
   including 480 octets in length.  All transport receiver
   implementations SHOULD be able to accept messages of up to and
   including 2048 octets in length.  Transport receivers MAY receive
   messages larger than 2048 octets in length.  If a transport receiver
   receives a message with a length larger than it supports, the
   transport receiver SHOULD truncate the payload.  Alternatively, it
   MAY discard the message.
>
>   If a transport receiver truncates messages, the truncation MUST occur
   at the end of the message.  After truncation, the message MAY contain
   invalid UTF-8 encoding or invalid STRUCTURED-DATA.  The transport
   receiver MAY discard the message or MAY try to process as much as
   possible in this case.

So there are minimum lengths that must be supported, but I'm sure modern day syslog implements support significantly larger than that. Syslog messages must fit in a single UDP packet, so that is likely the real upper bound, but that's like 65k bytes.

mbeckerle commented on README.md:
```diff
@@ -1,32 +1,50 @@
 
+## Change History
+
+### 1.1.0
```
You need to modify the build.sbt to create this 1.1.0 release as the version.

Once finalized, tag as 1.1.0, and create a github release for that tag, send an announcement email to users mailing list, etc.

mbeckerle commented on src/test/scala/com/owlcyberdefense/syslog/testsSyslogPortable.scala:
```diff
@@ -48,5 +48,5 @@ class TestSyslogPortable {
 
   import TestSyslogPortable._
 
```
So are you keeping this file as stub for in case someone re-introduces the IBM-portable flavor?

I find it is helpful to comment out only the "@Test" pragma, so scala compilers don't complain about unused imports.

stevedlawrence replied:
Yeah, it would be nice it we could eventually make this portable. I'll just comment out `@test`

mbeckerle commented on src/test/resources/com/owlcyberdefense/syslog/testsSyslogPortable.tdml:
I could not find a definition for element name="Syslog_Portable", not even commented out. 

If this file is being retained in case someone wants to re-enable the IBM DFDL variant, add comments to that effect.

mbeckerle reviewed (approved):
Looks good. It is great to have a generic syslog that handles all the variants. 

You should run an XSAT-style tool on this just to get ahead of those warnings that will otherwise come up in future deployments of this schema. 

I did not see build.sbt in the review, I think was not added?

Arguably, this is not really backward compatible so the version should be 2.0.0. The elements and their namespaces are somewhat different now I think. I may be wrong there. If this only adds a new universal syslog variant, then 1.1.0 is fine.

stevedlawrence committed:
Refactor to support all Syslog variants

- Move all Syslog variants into their own files, with root elements to
  support parsing only a single variant
- Modify the Syslog root so it supports all variants instead of just
  RFC5424
- Copy Solarwinds tests to Sylog tests with modifications to remove
  Solarwinds specific parts to ensure the Syslog element supports all
  variants
- Removes portability. PriorityValue is now referenced from multiple
  schemas so it is difficult to have a single "portable" schema.
  Instead, instructions and documentation are added to describe how make
  the schema portable.
- Resolve all XSAT findings

stevedlawrence committed:
Bump to version 2.0.0

- Update build.sbt
- Update release notes
